src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -29,27 +32,70 @@ namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
 
 internal abstract class LanguageServerProjectLoader
 {
-    protected readonly AsyncBatchingWorkQueue<ProjectToLoad> ProjectsToLoadAndReload;
+    private readonly AsyncBatchingWorkQueue<ProjectToLoad> _projectsToLoadAndReload;
 
     protected readonly ProjectSystemProjectFactory ProjectFactory;
     private readonly ProjectTargetFrameworkManager _targetFrameworkManager;
     private readonly ProjectSystemHostInfo _projectSystemHostInfo;
     private readonly IFileChangeWatcher _fileChangeWatcher;
-    private readonly IGlobalOptionService _globalOptionService;
+    protected readonly IGlobalOptionService GlobalOptionService;
     protected readonly ILoggerFactory LoggerFactory;
     private readonly ILogger _logger;
     private readonly ProjectLoadTelemetryReporter _projectLoadTelemetryReporter;
     private readonly BinlogNamer _binlogNamer;
-    private readonly ProjectFileExtensionRegistry _projectFileExtensionRegistry;
     protected readonly ImmutableDictionary<string, string> AdditionalProperties;
 
     /// <summary>
-    /// The list of loaded projects in the workspace, keyed by project file path. The outer dictionary is a concurrent dictionary since we may be loading
-    /// multiple projects at once; the key is a single List we just have a single thread processing any given project file. This is only to be used
-    /// in <see cref="LoadOrReloadProjectsAsync" /> and downstream calls; any other updating of this (like unloading projects) should be achieved by adding
-    /// things to the <see cref="ProjectsToLoadAndReload" />.
+    /// Guards access to <see cref="_loadedProjects"/>.
+    /// To keep the LSP queue responsive, <see cref="_gate"/> must not be held while performing design-time builds.
     /// </summary>
-    private readonly ConcurrentDictionary<string, List<LoadedProject>> _loadedProjects = [];
+    private readonly SemaphoreSlim _gate = new SemaphoreSlim(1);
+
+    /// <summary>
+    /// Guarded by <see cref="_gate"/>.
+    /// Maps the file path of a tracked project to the load state for the project.
+    /// Absence of an entry indicates the project is not tracked, e.g. it was never loaded, or it was unloaded.
+    /// <see cref="_gate"/> must be held when modifying the dictionary or objects contained in it.
+    /// </summary>
+    private readonly Dictionary<string, ProjectLoadState> _loadedProjects = [];
+
+    /// <summary>
+    /// State transitions:
+    /// <see cref="BeginLoading"/> -> <see cref="Loaded"/>
+    /// <see cref="BeginLoadingWithPrimordial"/> -> <see cref="Loaded"/>
+    /// Any state -> unloaded (which is denoted by removing the <see cref="_loadedProjects"/> entry for the project)
+    /// </summary>
```

Comment:
I don't mind the state concept here to make it clear that some can be in the primordial state and that's very different than having regular projects, but there's a lot of redundancy here that's creating duplication elsewhere:

1. We have two "WithPrimordial" that needs to be checked to see if it has an ID (which results in duplicate code in unload which is not good).
2. BeginLoading's handling in code is identical to Loaded with an empty list. Since a load can always mean we introduce new TFMs, it's not like there's any expectation that once we hit loaded we won't add or remove projects later.

Two states are all that's needed:

1. "Primordial" (with the ID)
2. "Regular' (or some other name) that's just the list of LoadedProjects.

Reply:
Done